File: apps/web/src/domain/descriptionParagraphs.ts
/**
 * Paragraph splitting for scenario descriptions.
 * Contract: packages/contracts/fixtures/description_paragraphs.json
 * Python mirror: apps/api/app/core/description_paragraphs.py
 */
export const PARAGRAPH_SEPARATOR = "\n\n";

export function splitDescriptionParagraphs(description: string): string[] {
  const text = (description || "").trim();
  if (!text) return [];
  if (text.includes(PARAGRAPH_SEPARATOR)) {
    return text.split(PARAGRAPH_SEPARATOR).filter((p) => p.trim());
  }
  return [text];
}

export function joinDescriptionParagraphs(parts: string[]): string {
  return parts.join(PARAGRAPH_SEPARATOR);
}

export function replaceDescriptionParagraph(
  description: string,
  paragraphIndex: number,
  newText: string,
): string {
  const parts = splitDescriptionParagraphs(description);
  if (paragraphIndex < 0 || paragraphIndex >= parts.length) {
    throw new Error(
      `paragraph_index ${paragraphIndex} out of range (0..${parts.length - 1})`,
    );
  }
  parts[paragraphIndex] = newText;
  return joinDescriptionParagraphs(parts);
}

File: apps/api/app/core/description_paragraphs.py
"""Split scenario descriptions into paragraphs (double newline separated).

Contract: packages/contracts/fixtures/description_paragraphs.json
Web mirror: apps/web/src/domain/descriptionParagraphs.ts
"""
from __future__ import annotations

PARAGRAPH_SEPARATOR = "\n\n"


def split_paragraphs(description: str) -> list[str]:
    text = (description or "").strip()
    if not text:
        return []
    if PARAGRAPH_SEPARATOR in text:
        return [p for p in text.split(PARAGRAPH_SEPARATOR) if p.strip()]
    return [text]


def paragraph_count(description: str) -> int:
    return len(split_paragraphs(description))


def replace_paragraph(*, description: str, paragraph_index: int, new_text: str) -> str:
    parts = split_paragraphs(description)
    if paragraph_index < 0 or paragraph_index >= len(parts):
        msg = f"paragraph_index {paragraph_index} out of range (0..{len(parts) - 1})"
        raise ValueError(msg)
    parts[paragraph_index] = new_text
    return PARAGRAPH_SEPARATOR.join(parts)
